src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -382,17 +382,164 @@ See Zulip thread:
+      - q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) :=
+begin
+  have h : (p • 1 + q • adjoin_root.mk W.polynomial X) * adjoin_root.mk W.polynomial X
+    = (q * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) • 1
+      + (p - q * (C W.a₁ * X + C W.a₃)) • adjoin_root.mk W.polynomial X,
+  { simp only [smul],
+    exact adjoin_root.mk_eq_mk.mpr
+      ⟨C q, by { simp only [weierstrass_curve.polynomial, C_sub, C_mul], ring1 }⟩ },
+  simp_rw [algebra.norm_eq_matrix_det W^.coordinate_ring.basis, matrix.det_fin_two,
+           algebra.left_mul_matrix_eq_repr_mul, basis_zero, mul_one, basis_one, h, map_add,
+           finsupp.add_apply, map_smul, finsupp.smul_apply, ← basis_zero, ← basis_one,
+           basis.repr_self_apply, if_pos, if_neg one_ne_zero, if_neg zero_ne_one, smul_eq_mul],
+  ring1
+end
+
+lemma nat_degree_norm_ne_one [nontrivial R] [no_zero_divisors R] (x : W.coordinate_ring) :
+  (algebra.norm R[X] x).nat_degree ≠ 1 :=
+begin
+  rcases exists_smul_basis_eq x with ⟨p, q, rfl⟩,
+  rw [norm_smul_basis],
+  by_cases hq : q = 0,
+  { rw [hq, mul_zero, zero_mul, sub_zero, zero_pow zero_lt_two, zero_mul, sub_zero, nat_degree_pow,
+        ← zero_add 1, ← mul_zero 2],
+    exact nat.two_mul_ne_two_mul_add_one },
+  { have hX : X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆ ≠ 0,
+    { convert_to cubic.to_poly ⟨1, W.a₂, W.a₄, W.a₆⟩ ≠ 0,
+      { simp only [cubic.to_poly, C_1], ring1 },
+      exact cubic.ne_zero_of_a_ne_zero one_ne_zero },
+    have hp : (p * q * (C W.a₁ * X + C W.a₃)).nat_degree ≤ p.nat_degree + q.nat_degree + 1,
+    { by_cases hp : p = 0,
+      { simp only [hp, zero_mul, nat_degree_zero, zero_le] },
+      { by_cases hq : q = 0,
+        { simp only [hq, mul_zero, zero_mul, nat_degree_zero, zero_le] },
+        { by_cases hX : C W.a₁ * X + C W.a₃ = 0,
+          { simp only [hX, mul_zero, nat_degree_zero, zero_le] },
+          { simp only [nat_degree_mul (mul_ne_zero hp hq) hX, nat_degree_mul hp hq,
+                       add_le_add_iff_left, nat_degree_linear_le] } } } },
+    have hq :
+      (q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)).nat_degree = 2 * q.nat_degree + 3,
+    { rw [nat_degree_mul (pow_ne_zero 2 hq) hX, nat_degree_pow, ← one_mul $ X ^ 3, ← C_1],
+      exact congr_arg _ (nat_degree_cubic one_ne_zero) },
+    by_cases h : p.nat_degree ≤ q.nat_degree + 1,
+    { rw [nat_degree_sub_eq_right_of_nat_degree_lt],
+      { linarith only [hq] },
+      { apply lt_of_le_of_lt (nat_degree_sub_le _ _) (max_lt _ _);
+          linarith only [nat_degree_pow p 2, hp, hq, h] } },
+    { rw [sub_sub, nat_degree_sub_eq_left_of_nat_degree_lt],
+      { rw [nat_degree_pow p 2, ← zero_add 1, ← mul_zero 2],
+        exact nat.two_mul_ne_two_mul_add_one },
+      { apply lt_of_le_of_lt (nat_degree_add_le _ _) (max_lt _ _);
+          linarith only [nat_degree_pow p 2, hp, hq, h] } } }
+end
+
+lemma degree_norm_ne_one [nontrivial R] [no_zero_divisors R] (x : W.coordinate_ring) :
```

Comment:
`[nontrivial R] [no_zero_divisors R]` is usually spelt `[is_domain R]`, because they are equivalent and the latter is shorter, and having a single class probably saves typeclass inference efforts, even though the latter is "more general" as typeclass inference can infer the former two from the latter but not vice versa.